module/apmslog/handler.go

```diff
@@ -0,0 +1,173 @@
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+package apmslog // import "go.elastic.co/apm/module/apmslog/v2"
+
+import (
+	"context"
+	"errors"
+
+	"log/slog"
+	"slices"
+	"strings"
+
+	"go.elastic.co/apm/v2"
+)
+
+const (
+	// FieldKeyTraceID is the field key for the trace ID.
+	FieldKeyTraceID = "trace.id"
+
+	// FieldKeyTransactionID is the field key for the transaction ID.
+	FieldKeyTransactionID = "transaction.id"
+
+	// FieldKeySpanID is the field key for the span ID.
+	FieldKeySpanID = "span.id"
+)
+
+type ApmHandler struct {
+	Tracer       *apm.Tracer
+	ReportLevels []slog.Level
+	Handler      slog.Handler
+}
+
+func (s *ApmHandler) tracer() *apm.Tracer {
+	if s.Tracer == nil {
+		return apm.DefaultTracer()
+	}
+	return s.Tracer
+}
+
+func (s *ApmHandler) levels() []slog.Level {
+	if s.ReportLevels == nil {
+		return []slog.Level{slog.LevelError}
+	}
+	return s.ReportLevels
+}
+
+// Enabled reports whether the handler handles records at the given level.
+func (s *ApmHandler) Enabled(ctx context.Context, level slog.Level) bool {
+	return s.Handler.Enabled(ctx, level)
+}
+
+// WithAttrs returns a new ApmHandler with passed attributes attached.
+func (s *ApmHandler) WithAttrs(attrs []slog.Attr) slog.Handler {
+	return &ApmHandler{s.Tracer, s.ReportLevels, s.Handler.WithAttrs(attrs)}
+}
+
+// WithGroup returns a new ApmHandler with passed group attached.
+func (s *ApmHandler) WithGroup(name string) slog.Handler {
+	return &ApmHandler{s.Tracer, s.ReportLevels, s.Handler.WithGroup(name)}
+}
+
+func (s *ApmHandler) Handle(ctx context.Context, r slog.Record) error {
+
+	// report record as APM error
+	tracer := s.tracer()
+	if slices.Contains(s.levels(), r.Level) && tracer.Recording() {
+
+		// attempt to find error/err attribute
+		// slog doesnt have a standard way of attaching an
+		// error to a record, so attempting to grab any attribute
+		// that has error/err as key and extracting the value
+		// seems like a likely way to do it.
+		var err error
+		r.Attrs(func(a slog.Attr) bool {
+			if a.Key == "error" || a.Key == "err" {
+				if v, ok := a.Value.Any().(error); ok {
+					err = v
+					return false
+				}
+				if v, ok := a.Value.Any().(string); ok {
+					err = errors.New(v)
+					return false
+				}
+				return false
+			}
+			return true
+		})
+		// if error/err attribute exists, use it as Error value
+		var errLogRecord apm.ErrorLogRecord
+		if err != nil {
+			errLogRecord = apm.ErrorLogRecord{
+				Message: r.Message,
+				Level:   strings.ToLower(r.Level.String()),
+				Error:   err,
+			}
+		} else {
+			errLogRecord = apm.ErrorLogRecord{
+				Message: r.Message,
+				Level:   strings.ToLower(r.Level.String()),
+			}
+		}
```

Comment:
This can be simplified:

```suggestion
		errLogRecord := apm.ErrorLogRecord{
			Message: r.Message,
			Level:   strings.ToLower(r.Level.String()),
			Err: err,
		}
```

If `err` is nil, `errLogRecord.Err` will also be `nil`, which is the default value anyway.

Reply:
You then don't even need to have an `errLogRecord` variable, you can just create `errlog`.

See the logrus implementation: https://github.com/elastic/apm-agent-go/blob/d6810ab0a60bfaaed9f35f98933877ebbfb3803c/module/apmlogrus/hook.go#L95

Reply:
good catch!